### PR TITLE
Feature/csv table

### LINKED ISSUE #499: CSV Renderer
It would be great to copy / paste CSV (or any table-based format) and render it as a table.

For a reference on how it could work, that's done by the HackMD version of CodiMD in release 2.2.0

It could be
```csvpreview {header=true}
firstName, lastName, Phone
John, A, 1234
Jane, B, 4321
```

### PR DESCRIPTION
**Pull request by DerMolly**
### Component/Part
markdown render

### Description
This PR adds a CSV Replacer.

### Steps

- [x] added implementation
- [ ] added / updated tests
- [ ] added / updated documentation
- [x] extended changelog

### Related Issue(s)
closes #499

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -37,6 +37,7 @@
 - The Toolbar includes an EmojiPicker
 - Added shortcodes for [fork-awesome icons](https://forkaweso.me/Fork-Awesome/icons/) (e.g. `:fa-picture-o:`)
 - The code button now adds code fences even if the user selected nothing beforehand
+- Code fences with 'csv' as the language will now be rendered as tables.
```
Comment (mrdrogdrog) [inline CHANGELOG.md:40]: `Code blocks with 'csv' as language render as tables.`

File: src/components/markdown-renderer/replace-components/csv/csv-replacer.tsx [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+export class CsvReplacer implements ComponentReplacer {
+  getReplacement (codeNode: DomElement, index: number): React.ReactElement | undefined {
+    if (codeNode.name !== 'code' || !codeNode.attribs || !codeNode.attribs['data-highlight-language'] || codeNode.attribs['data-highlight-language'] !== 'csv' || !codeNode.children || !codeNode.children[0]) {
+      return
+    }
+
+    const code = codeNode.children[0].data as string
+
+    const extraData = codeNode.attribs['data-extra']
+    const extraInfos = /\s*(delimiter=([^\s]*))?\s*(header)?/.exec(extraData)
```
Comment (mrdrogdrog) [inline src/components/markdown-renderer/replace-components/csv/csv-replacer.tsx:15]: Please extract this regex into a const outside of the function.

File: src/components/markdown-renderer/replace-components/csv/csv-table.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+const parseCsvToRowsAndColumn = (csvText: string, csvColumnDelimiter: string): string[][] => {
+  const rows = csvText.split('\n');
+  if (!rows || rows.length === 0) {
+    return [];
+  }
+
+  return rows.map(row => row.split(csvColumnDelimiter));
+}
+
+export const CsvTable: React.FC<CsvTableProps> = ({ index, code, delimiter, showHeader, tableRowClassName, tableColumnClassName}) => {
```
Comment (mrdrogdrog) [inline src/components/markdown-renderer/replace-components/csv/csv-table.tsx:21]: You're calculating the CSV rows and columns on every render, even if the props didn't change. Please add some useMemo-Calls to cache results.

File: src/components/markdown-renderer/replace-components/highlighted-fence/highlighted-fence-replacer.tsx [hunk 4]
```diff
@@ -12,18 +13,31 @@ export class HighlightedCodeReplacer implements ComponentReplacer {
...
 
     if (showLineNumbers) {
       this.lastLineNumber = startLineNumber + code.split('\n')
         .filter(line => !!line).length
     }
 
+    if (language === 'csv') {
+      return <CsvToHtmlTable data={code} csvDelimiter=";"/>
+    }
+
```
Comment (mrdrogdrog) [inline src/components/markdown-renderer/replace-components/highlighted-fence/highlighted-fence-replacer.tsx:40]: I think you forgot to remove this.

**Review by mrdrogdrog — CHANGES_REQUESTED** [review]

File: src/components/markdown-renderer/replace-components/csv/csv-replacer.tsx (hunk 2 above)
Comment (DerMolly) [reply src/components/markdown-renderer/replace-components/csv/csv-replacer.tsx:15]: done

File: CHANGELOG.md (hunk 1 above)
Comment (DerMolly) [reply CHANGELOG.md:40]: done

File: src/components/markdown-renderer/replace-components/csv/csv-table.tsx (hunk 3 above)
Comment (DerMolly) [reply src/components/markdown-renderer/replace-components/csv/csv-table.tsx:21]: done

File: src/components/markdown-renderer/replace-components/highlighted-fence/highlighted-fence-replacer.tsx (hunk 4 above)
Comment (DerMolly) [reply src/components/markdown-renderer/replace-components/highlighted-fence/highlighted-fence-replacer.tsx:40]: oh yeah
fixed

File: src/components/markdown-renderer/replace-components/csv/csv-table.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+  const renderTableBody = (rows: string[][]) => {
+    return (
+      <tbody>
+      {
+        rows.map((row, rowNumber) => (
+          <tr className={tableRowClassName} key={`${index}-row-${rowNumber}`}>
+            {
+              row.map && row.map((column, columnIndex) => (
+                <td
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/replace-components/csv/csv-table.tsx:59]:
```suggestion
              row.map((column, columnIndex) => (
```

File: src/components/markdown-renderer/replace-components/csv/csv-table.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+                  className={tableColumnClassName}
+                  key={`${index}-cell-${rowNumber}${columnIndex}`}
+                >
```
Comment (ErikMichelson) [inline src/components/markdown-renderer/replace-components/csv/csv-table.tsx:62]:
```suggestion
                  key={`${index}-cell-${rowNumber}-${columnIndex}`}
                >
```

If you have row 11, column 1 and row 1, column 11, they would have the same index.

**Review by ErikMichelson — CHANGES_REQUESTED** [review]

**Review by mrdrogdrog — APPROVED** [review]

**Review by ErikMichelson — APPROVED** [review]

**Pull request merged**